pkg/agent/httpmux.go

```diff
@@ -0,0 +1,117 @@
+package agent
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+
+	"go.uber.org/zap"
+)
+
+// HttpMuxServer provides a way to multiplex multiple http.ServeMux
+// instances on a single HTTP server, so that requests are routed to
+// the correct ServeMux based on a prefix at the beginning of the URL
+// path.
+//
+// For example, if you call RegisterMux("foo", myMux), then all requests to
+// "/foo/*" are routed to 'myMux' instance
+//
+// (Why is this needed? You could register all the routes directly in
+// one giant http.ServeMux, but http.ServeMux doesn't provide any way
+// to unregister patterns.)
+type HttpMuxServer struct {
+	muxedServers map[string]*http.ServeMux
+	lock         sync.Mutex // protects muxedServers
+	logger       *zap.Logger
+}
+
+// Implements http.Handler
+func (s *HttpMuxServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	s.logger.Debug(fmt.Sprintf("MUX: received request: %s", r.URL.Path))
+
+	path, found := strings.CutPrefix(r.URL.Path, "/")
+	if !found {
+		w.WriteHeader(http.StatusBadRequest)
+		_, _ = w.Write([]byte("missing /"))
+		return
+	}
+
+	muxID, path, found := strings.Cut(path, "/")
+	if !found {
+		w.WriteHeader(http.StatusBadRequest)
+		_, _ = w.Write([]byte("request must start with mux ID"))
+		return
+	}
+
+	s.lock.Lock()
+	subServer := s.muxedServers[muxID]
+	s.lock.Unlock()
+	if subServer == nil {
```

Comment:
```suggestion
	subServer, ok := s.muxedServers[muxID]
	s.lock.Unlock()
	if !ok {
```

Generally preferable to explicitly check whether the value exists, rather than setting it to nil. There's a couple other places that do this as well.